benchmarks/bench_kn_nlearners.py

```diff
@@ -0,0 +1,311 @@
+    threshold=0.5,
+):
+    trueskill.setup(
+        mu=0.0,
+        sigma=np.sqrt(1e-9),
+        beta=float(np.sqrt(beta_sqr)),
+        tau=tau,
+        draw_probability=0.0,
+        backend="mpmath",
+    )
+
+    user_model = {"mean": {}, "variance": {}}
+
+    predicted = []
+
+    for _, (topic_dict, label) in enumerate(records):
+        # create_teams
+        team_learner = tuple()
+        team_learner_mean_vec = []
+        team_learner_var_vec = []
+
+        team_content = tuple()
+        team_content_mean_vec = []
+        team_content_var_vec = []
+
+        topic_seq = []
+
+        for topic, coverage in topic_dict.items():
+            topic_seq.append(topic)
+            # get user skill rating
+            tmp_learner_mean = user_model["mean"].get(topic, init_skill)
+            tmp_learner_var = user_model["variance"].get(topic, def_var)
+            learner_skill = trueskill.Rating(
+                mu=float(tmp_learner_mean), sigma=float(np.sqrt(tmp_learner_var))
+            )
+
+            team_learner += (learner_skill,)
+            team_learner_mean_vec.append(tmp_learner_mean)
+            team_learner_var_vec.append(tmp_learner_var)
+
+            # get skill coverage
+            tmp_coverage = coverage
+            tmp_content_var = 1e-9
+            topic_cov = trueskill.Rating(
+                mu=tmp_coverage, sigma=float(np.sqrt(tmp_content_var))
+            )
+            team_content += (topic_cov,)
+            team_content_mean_vec.append(tmp_coverage)
+            team_content_var_vec.append(tmp_content_var)
+
+        # check if user engages
+        pred_prob = team_sum_quality(
+            team_learner_mean_vec,
+            team_learner_var_vec,
+            team_content_mean_vec,
+            team_content_var_vec,
+            beta_sqr,
+        )
+
+        prediction = int(pred_prob >= threshold)
+
+        # if label is negative and setting is positive only, skip updating
+        if positive_only and label != 1:
+            pass
+        else:
+            if label == 1:
+                # learner wins
+                new_team_learner, _ = trueskill.rate(
+                    [team_learner, team_content], ranks=[0, 1]
+                )
+            else:
+                # content wins
+                _, new_team_learner = trueskill.rate(
+                    [team_content, team_learner], ranks=[0, 1]
+                )
+
+            for _idx, topic in enumerate(topic_seq):
+                user_model["mean"][topic], user_model["variance"][topic] = (
+                    new_team_learner[_idx].mu,
+                    new_team_learner[_idx].sigma ** 2,
+                )
+
+        predicted.append(prediction)
+
+
+def truelearn_novel(
+    records,
+    positive_only=False,
+    init_skill=0.0,
+    def_var=0.5,
+    tau=0.0,
+    beta_sqr=0.35**2,
+    threshold=0.5,
+    draw_probability="dynamic",
+    draw_factor=0.1,
+):
+    user_model = {"mean": {}, "variance": {}}
```

Comment:
I see we use this to represent our user model, any reason why we dont use our own user model?

Reply:
This is because this is a benchmark. The code here is from Sahan's experiments.